plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -378,190 +435,194 @@ extendEnv :: Name -> CekValue uni fun -> CekValEnv uni fun -> CekValEnv uni fun
+          TypeArg:arity' ->
+              case arity' of
+                [] -> applyBuiltin ctx bn args  -- Final argument is a type argument
+                _  -> returnCek ctx $ VBuiltin ex bn arity0 arity' (forces + 1) args -- More arguments expected
+    forceEvaluate _ val =
+            throwingDischarged _MachineError NonPolymorphicInstantiationMachineError val
+
+    -- | Apply a function to an argument and proceed.
+    -- If the function is a lambda 'lam x ty body' then extend the environment with a binding of @v@
+    -- to x@ and call 'computeCek' on the body.
+    -- If the function is a builtin application then check that it's expecting a term argument, and if
+    -- it's the final argument then apply the builtin to its arguments, return the result, or extend
+    -- the value with the new argument and call 'returnCek'. If v is anything else, fail.
+    applyEvaluate
+        :: Context uni fun
+        -> CekValue uni fun   -- lhs of application
+        -> CekValue uni fun   -- rhs of application
+        -> CekM cost uni fun s (Term Name uni fun ())
+    applyEvaluate ctx (VLamAbs _ name body env) arg =
+        computeCek ctx (extendEnv name arg env) body
+    applyEvaluate ctx val@(VBuiltin ex bn arity0 arity forces args) arg = do
+        case arity of
+          []        -> throwingDischarged _MachineError EmptyBuiltinArityMachineError val
+                    -- Should be impossible: see forceEvaluate.
+          TypeArg:_ -> throwingDischarged _MachineError UnexpectedBuiltinTermArgumentMachineError val'
+                       where val' = VBuiltin ex bn arity0 arity forces (args++[arg]) -- reconstruct the bad application
+          TermArg:arity' -> do
+              let args' = args ++ [arg]
+              case arity' of
+                [] -> applyBuiltin ctx bn args' -- 'arg' was the final argument
+                _  -> returnCek ctx $ VBuiltin ex bn arity0 arity' forces args'  -- More arguments expected
+    applyEvaluate _ val _ = throwingDischarged _MachineError NonFunctionalApplicationMachineError val
+
+    -- | Apply a builtin to a list of CekValue arguments
+    applyBuiltin
+        :: Context uni fun
+        -> fun
+        -> [CekValue uni fun]
+        -> CekM cost uni fun s (Term Name uni fun ())
+    applyBuiltin ctx bn args = do
+      rt <- asks cekEnvRuntime
+      BuiltinRuntime sch _ f exF <- lookupBuiltinExc (Proxy @(CekEvaluationException uni fun)) bn rt
+
+      -- ''applyTypeSchemed' doesn't throw exceptions so that we can easily catch them here and
+      -- post-process them.
+      -- See Note [Being generic over @term@ in 'CekM'].
+      resultOrErr <- runExceptT $ applyTypeSchemed bn sch f exF args
+      case resultOrErr of
+          -- Turn the cause of a possible failure, being a 'CekValue', into a 'Term'.
+          Left e       -> throwCek $ mapCauseInMachineException (void . dischargeCekValue) e
```

Comment:
Oh, hm, I see how you manage to avoid being generic over `term` now. And yeah, this reasoning also makes sense:

> The main reason for this is that we have some awkwardness at
> the interface of `applyTypeSchemed` where we need to change the cause
> part of the error from a value into a term, and I don't want to have to
> catch and rethrow an exception there if we can avoid it.

Well, maybe there's something we can do about it, maybe not. Let's merge the PR and I'll think about the builtin application machinery later.

Reply:
Yeah, I think it might be most natural to get rid of the need to change it, by carrying values in errors. That would leak them out, but it would also stop us doing discharging when making errors.